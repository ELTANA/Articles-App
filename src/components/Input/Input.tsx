import { FC, forwardRef } from 'react';
import type { InputProps } from './Input.types';
import useInputError from '$hooks/useInputError';
import Search from '$svgs/search';

const Input: FC<InputProps> = forwardRef<HTMLDivElement, InputProps>(
  ({ error, type, label, name, inputRef, ...rest }, ref) => {
    const { errorComponent } = useInputError({ error });
    return (
      <div ref={ref} className="flex flex-col gap-4 w-full">
        {label ? (
          <label className="text-sm md:text-base text-gray-600 capitalize font-medium" htmlFor={name}>
            {label}
          </label>
        ) : null}
        <div className="w-full relative h-max">
          <input
            ref={inputRef}
            type={type}
            className={`w-full shadow-sm bg-white rounded border text-base outline-none text-gray-700 py-1 px-3 leading-8 transition-colors duration-200 ease-in-out ${error ? 'border-red-300 focus:border-red-300 focus:ring-2 focus:ring-red-300' : 'border-blue-100 focus:border-blue-300 focus:ring-2 focus:ring-blue-200'}`}
            aria-label={label}
            {...rest}
            name={name}
            id={name}
            // {...register}
          />
          {type === 'search' ? (
            <button type="submit" className="absolute top-1/2 -translate-y-1/2 right-[6px]" role="searchbox">
              <Search />
            </button>
          ) : null}
        </div>
        {errorComponent}
      </div>
    );
  }
);

Input.displayName = 'Input';

export default Input;
