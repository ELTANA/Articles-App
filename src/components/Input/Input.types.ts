import { InputHTMLAttributes } from 'react';
import type { FieldError } from 'react-hook-form';

export interface IInputBaseType {
  label?: string;
  error?: FieldError;
  name: string;
}

export type InputProps = IInputBaseType & InputHTMLAttributes<HTMLInputElement>;

export type TextAreaProps = IInputBaseType & InputHTMLAttributes<HTMLTextAreaElement>;
