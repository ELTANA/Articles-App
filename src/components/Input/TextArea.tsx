import { FC, forwardRef } from 'react';
import type { TextAreaProps } from './Input.types';
import useInputError from '$hooks/useInputError';

const TextArea: FC<TextAreaProps> = forwardRef<HTMLDivElement, TextAreaProps>(
  ({ error, label, name, textAreaRef, ...rest }, ref) => {
    const { errorComponent } = useInputError({ error });
    return (
      <div ref={ref} className="flex flex-col gap-4 w-full">
        {label ? (
          <label className="text-sm md:text-base text-gray-600 capitalize font-medium" htmlFor={name}>
            {label}
          </label>
        ) : null}
        <div className="w-full relative h-max">
          <textarea
            ref={textAreaRef}
            name={name}
            id={name}
            className={`w-full shadow-sm bg-white rounded border text-base outline-none text-gray-700 py-1 px-3 leading-8 transition-colors duration-200 ease-in-out resize-none h-[150px] ${error ? 'border-red-200 focus:border-red-300 focus:ring-2 focus:ring-red-300' : 'border-blue-200 focus:border-blue-100 focus:ring-2 focus:ring-blue-100'}`}
            aria-label={label}
            {...rest}
          ></textarea>
        </div>
        {errorComponent}
      </div>
    );
  }
);

TextArea.displayName = 'TextArea';

export default TextArea;
