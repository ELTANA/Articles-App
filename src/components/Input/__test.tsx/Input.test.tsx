import { render, screen, waitFor } from '@testing-library/react';
import Input from '../Input';
import '@testing-library/jest-dom';
import { userEvent } from '@testing-library/user-event';

beforeAll(() => {
  window.ResizeObserver =
    window.ResizeObserver ||
    jest.fn().mockImplementation(() => ({
      disconnect: jest.fn(),
      observe: jest.fn(),
      unobserve: jest.fn()
    }));
});

describe('Input component', () => {
  test('renders Input with label and without error', () => {
    render(<Input label="Test Label" name="testInput" type="text" />);

    const labelElement = screen.getByText(/Test Label/i);
    const inputElement = screen.getByRole('textbox', { name: 'Test Label' });

    expect(labelElement).toBeInTheDocument();
    expect(inputElement).toBeInTheDocument();
  });

  test('renders Input with error', () => {
    render(<Input label="Test Label" name="testInput" type="text" error={{ type: 'value', message: 'Test error' }} />);

    const errorElement = screen.getByText(/Test error/i);

    expect(errorElement).toBeInTheDocument();
  });

  test('handles search input type', () => {
    render(<Input label="Search" name="searchInput" type="search" role="searchbox" />);

    const searchButton = screen.getByRole('searchbox', { name: 'Search' });

    expect(searchButton).toBeInTheDocument();
  });

  // test('handles user input', () => {
  //   const handleChange = jest.fn();
  //   render(<Input label="Test Label" name="testInput" type="text" onChange={handleChange} />);

  //   const inputElement = screen.getByRole('textbox', { name: 'Test Label' });

  //   userEvent.type(inputElement, 'Hello');

  //   expect(handleChange).toHaveBeenCalledTimes(5); // Each keystroke triggers onChange
  //   expect(inputElement).toHaveValue('Hello');
  // });

  test('handles user input', async () => {
    const handleChange = jest.fn();
    render(<Input label="Test Label" name="testInput" type="text" onChange={handleChange} />);

    const inputElement = screen.getByRole('textbox', { name: 'Test Label' });

    userEvent.type(inputElement, 'Hello');

    // Use waitFor to wait for asynchronous changes
    await waitFor(() => {
      // Expect handleChange to be called for each character in 'Hello'
      expect(handleChange).toHaveBeenCalledTimes(5);
    });

    // You can also assert the value of the input
    expect(inputElement).toHaveValue('Hello');
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<Input label="Test Label" name="testInput" type="text" />);

    expect(asFragment()).toMatchSnapshot();
  });
});
