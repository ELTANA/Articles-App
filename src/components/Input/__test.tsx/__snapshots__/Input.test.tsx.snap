// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-4 w-full"
  >
    <label
      class="text-sm md:text-base text-gray-600 capitalize font-medium"
      for="testInput"
    >
      Test Label
    </label>
    <div
      class="w-full relative h-max"
    >
      <input
        aria-label="Test Label"
        class="w-full shadow-sm bg-white rounded border text-base outline-none text-gray-700 py-1 px-3 leading-8 transition-colors duration-200 ease-in-out border-blue-100 focus:border-blue-300 focus:ring-2 focus:ring-blue-200"
        id="testInput"
        name="testInput"
        type="text"
      />
    </div>
    <div
      class="h-fit transition-all"
    >
      <p
        class="font-normal text-sm leading-4 text-red-600 transition-all"
      />
    </div>
  </div>
</DocumentFragment>
`;
